src/clock_control/watchdog.rs

```diff
@@ -0,0 +1,227 @@
+#[derive(Debug)]
+pub struct WatchdogConfig<
+    T1: Into<MicroSeconds>,
+    T2: Into<MicroSeconds>,
+    T3: Into<MicroSeconds>,
+    T4: Into<MicroSeconds>,
+> {
+    pub period1: T1,
+    pub action1: WatchdogAction,
+    pub period2: T2,
+    pub action2: WatchdogAction,
+    pub period3: T3,
+    pub action3: WatchdogAction,
+    pub period4: T4,
+    pub action4: WatchdogAction,
+    pub cpu_reset_duration: WatchDogResetDuration,
+    pub sys_reset_duration: WatchDogResetDuration,
+    pub pause_in_sleep: bool,
+    pub reset_cpu: [bool; 2],
+}
+
+impl WatchDog {
+    /// internal function to create new watchdog structure
+    pub(crate) fn new(clock_control_config: super::ClockControlConfig) -> Self {
+        WatchDog {
+            clock_control_config,
+        }
+    }
+
+    /// function to unlock the watchdog (write unblock key) and lock after use
+    fn access_registers<A, F: FnMut(&esp32::rtccntl::RegisterBlock) -> A>(
+        &mut self,
+        mut f: F,
+    ) -> A {
+        // Unprotect write access to registers
+        let rtc_control = unsafe { &(*RTCCNTL::ptr()) };
+
+        rtc_control
+            .wdtwprotect
+            .write(|w| unsafe { w.bits(WATCHDOG_UNBLOCK_KEY) });
+
+        let a = f(rtc_control);
+
+        // Protect again
+        rtc_control
+            .wdtwprotect
+            .write(|w| unsafe { w.bits(WATCHDOG_BLOCK_VALUE) });
+
+        a
+    }
+
+    /// Calculate period from ref ticks
+    fn calc_period(&self, value: u32) -> MicroSeconds {
+        (((1000000u64 * value as u64)
+            / (u32::from(self.clock_control_config.slow_rtc_frequency()) as u64)) as u32)
+            .us()
+    }
+
+    /// Calculate ref ticks from period
+    fn calc_ticks(&self, value: MicroSeconds) -> u32 {
+        (u32::from(value) as u64 * u32::from(self.clock_control_config.slow_rtc_frequency()) as u64
+            / 1000000u64) as u32
+    }
+
+    /// Get watchdog configuration
+    pub fn config(
+        &self,
+    ) -> Result<WatchdogConfig<MicroSeconds, MicroSeconds, MicroSeconds, MicroSeconds>, super::Error>
+    {
+        let rtc_control = unsafe { &(*RTCCNTL::ptr()) };
+        let wdtconfig0 = rtc_control.wdtconfig0.read();
+
+        let stg0 = match wdtconfig0.wdt_stg0().variant() {
+            Val(x) => x,
+            _ => return Err(super::Error::UnsupportedWatchdogConfig),
+        };
+        let stg1 = match wdtconfig0.wdt_stg1().variant() {
+            Val(x) => x,
+            _ => return Err(super::Error::UnsupportedWatchdogConfig),
+        };
+        let stg2 = match wdtconfig0.wdt_stg2().variant() {
+            Val(x) => x,
+            _ => return Err(super::Error::UnsupportedWatchdogConfig),
+        };
+        let stg3 = match wdtconfig0.wdt_stg3().variant() {
+            Val(x) => x,
+            _ => return Err(super::Error::UnsupportedWatchdogConfig),
+        };
+
+        Ok(
+            WatchdogConfig::<MicroSeconds, MicroSeconds, MicroSeconds, MicroSeconds> {
```

Comment:
```suggestion
            WatchdogConfig {
```

The compiler can infer it, or not?

Reply:
Will try.

Reply:
Done.